File: app/assets/javascripts/dashboard/PortfolioController.js
(function () {
    var app = angular.module('shocktrade');

    /**
     * Portfolio Controller
     * @author lawrence.daniels@gmail.com
     */
    app.controller('PortfolioController', ['$scope', '$cookieStore', '$log', '$timeout', 'toaster', 'MySession', 'ContestService', 'NewOrderDialog', 'QuoteService',
        function ($scope, $cookieStore, $log, $timeout, toaster, MySession, ContestService, NewOrderDialog, QuoteService) {

            $scope.selectedClosedOrder = null;
            $scope.selectedOrder = null;
            $scope.selectedPosition = null;

            $scope.portfolioTabs = [{
                "name": "Chat",
                "icon": "fa-comment-o",
                "path": "/assets/views/dashboard/chat.htm",
                "active": false,
                "init": function () {
                },
                "isLocked": function () {
                    return false;
                }
            }, {
                "name": "Positions",
                "icon": "fa-list-alt",
                "path": "/assets/views/dashboard/positions.htm",
                "active": false,
                "init": function () {
                    // TODO enrich positions here
                },
                "isLocked": function () {
                    return MySession.getContestStatus() !== 'ACTIVE';
                }
            }, {
                "name": "Open Orders",
                "icon": "fa-folder-open-o",
                "path": "/assets/views/dashboard/orders_active.htm",
                "active": false,
                "init": function () {
                    // TODO enrich orders here
                    angular.forEach($scope.getActiveOrders(), function(o) {
                        $log.info("portfolioTabs: order => " + angular.toJson(o));
                    });
                },
                "isLocked": function () {
                    return MySession.getContestStatus() !== 'ACTIVE';
                }
            }, {
                "name": "Closed Orders",
                "icon": "fa-folder-o",
                "path": "/assets/views/dashboard/orders_closed.htm",
                "active": false,
                "init": function () {
                },
                "isLocked": function () {
                    return false;
                }
            }, {
                "name": "Performance",
                "icon": "fa-bar-chart-o",
                "path": "/assets/views/dashboard/performance.htm",
                "active": false,
                "init": function () {
                },
                "isLocked": function () {
                    return false;
                }
            }, {
                "name": "Exposure",
                "icon": "fa-pie-chart",
                "path": "/assets/views/dashboard/exposure.htm",
                "active": false,
                "init": function () {
                },
                "isLocked": function () {
                    return MySession.getContestStatus() !== 'ACTIVE';
                }
            }];

            /////////////////////////////////////////////////////////////////////
            //          Active Order Functions
            /////////////////////////////////////////////////////////////////////

            $scope.getActiveOrders = function () {
                enrichOrders(MySession.getParticipant());
                var orders = MySession.getOrders() || [];
                return orders.filter(function(o) {
                    return o.accountType === $scope.getAccountType();
                });
            };

            $scope.cancelOrder = function (contestId, playerId, orderId) {
                ContestService.deleteOrder(contestId, playerId, orderId)
                    .success(function (contest) {
                        MySession.setContest(contest);
                    })
                    .error(function (err) {
                        toaster.pop('error', 'Error!', "Failed to cancel order");
                    });
            };

            $scope.isMarketOrder = function (order) {
                return order.priceType == 'MARKET' || order.priceType == 'MARKET_ON_CLOSE';
            };

            $scope.isOrderSelected = function () {
                return $scope.getActiveOrders().length && $scope.selectedOrder != null;
            };

            $scope.selectOrder = function (order) {
                $scope.selectedOrder = order;
            };

            $scope.toggleSelectedOrder = function () {
                $scope.selectedOrder = null;
            };

            $scope.popupNewOrderDialog = function (params) {
                // were the parameters passed?
                params = params || {};

                var symbol = $cookieStore.get('symbol');
                if (!symbol) symbol = QuoteService.lastSymbol;
                params.symbol = symbol;

                NewOrderDialog.popup(params);
            };

            /////////////////////////////////////////////////////////////////////
            //          Closed Order Functions
            /////////////////////////////////////////////////////////////////////

            $scope.getClosedOrders = function () {
                var closedOrders = MySession.getOrderHistory() || [];
                return closedOrders.filter(function(o) {
                    return $scope.getClosedOrders.length && o.accountType === $scope.getAccountType();
                });
            };

            $scope.isClosedOrderSelected = function () {
                return $scope.selectedClosedOrder != null;
            };

            $scope.selectClosedOrder = function (closeOrder) {
                $scope.selectedClosedOrder = closeOrder;
            };

            $scope.toggleSelectedClosedOrder = function () {
                $scope.selectedClosedOrder = null;
            };

            $scope.orderCost = function (o) {
                return o.price * o.quantity + o.commission;
            };

            /////////////////////////////////////////////////////////////////////
            //          Performance Functions
            /////////////////////////////////////////////////////////////////////

            $scope.getPerformances = function () {
                return MySession.getPerformance() || [];
            };

            $scope.isPerformanceSelected = function () {
                return $scope.getPerformances().length && $scope.selectedPerformance != null;
            };

            $scope.selectPerformance = function (performance) {
                $scope.selectedPerformance = performance;
            };

            $scope.toggleSelectedPerformance = function () {
                $scope.selectedPerformance = null;
            };

            $scope.cost = function (tx) {
                return tx.pricePaid * tx.quantity + tx.commissions;
            };

            $scope.soldValue = function (tx) {
                return tx.priceSold * tx.quantity;
            };

            $scope.proceeds = function (tx) {
                return $scope.soldValue(tx) - $scope.cost(tx);
            };

            $scope.gainLoss = function (tx) {
                return 100.0 * ($scope.proceeds(tx) / $scope.cost(tx));
            };

            /////////////////////////////////////////////////////////////////////
            //          Position Functions
            /////////////////////////////////////////////////////////////////////

            $scope.getPositions = function () {
                enrichPositions(MySession.getParticipant());
                var positions = MySession.getPositions() || [];
                return positions.filter(function(p) {
                    return p.accountType === $scope.getAccountType();
                });
            };

            $scope.isPositionSelected = function () {
                return $scope.getPositions().length && $scope.selectedPosition != null;
            };

            $scope.selectPosition = function (position) {
                $scope.selectedPosition = position;
            };

            $scope.sellPosition = function (symbol, quantity) {
                NewOrderDialog.popup({symbol: symbol, quantity: quantity});
            };

            $scope.toggleSelectedPosition = function () {
                $scope.selectedPosition = null;
            };

            $scope.tradingStart = function () {
                return (new Date()).getTime(); // TODO replace with service call
            };

            /////////////////////////////////////////////////////////////////////
            //          Private Functions
            /////////////////////////////////////////////////////////////////////

            function enrichOrders(participant) {
                if (!MySession.participantIsEmpty()) {
                    if (!participant.enrichedOrders) {
                        participant.enrichedOrders = true;
                        ContestService.getEnrichedOrders(MySession.getContestID(), participant.OID())
                            .success(function (enrichedOrders) {
                                MySession.getParticipant().orders = enrichedOrders;
                            })
                            .error(function (err) {
                                toaster.pop('error', 'Error!', "Error loading enriched orders");
                            });
                    }
                }
            }

            function enrichPositions(participant) {
                if (!MySession.participantIsEmpty()) {
                    if (!participant.enrichedPositions) {
                        participant.enrichedPositions = true;
                        ContestService.getEnrichedPositions(MySession.getContestID(), participant.OID())
                            .success(function (enrichedPositions) {
                                MySession.getParticipant().positions = enrichedPositions;
                            })
                            .error(function (err) {
                                toaster.pop('error', 'Error!', "Error loading enriched positions");
                            });
                    }
                }
            }

            function resetOrders() {
                $scope.selectedOrder = null;
                $scope.selectedClosedOrder = null;
            }

            function resetPositions() {
                $scope.selectedPosition = null;
                $scope.selectedPerformance = null;
            }

            //////////////////////////////////////////////////////////////////////
            //              Watch Event Listeners
            //////////////////////////////////////////////////////////////////////

            $scope.$on("contest_selected", function (event, contest) {
                $log.info("[Portfolio] Contest '" + contest.name + "' selected");
                resetOrders();
                resetPositions();
            });

            $scope.$on("contest_updated", function (event, contest) {
                $log.info("[Portfolio] Contest '" + contest.name + "' updated");
                resetOrders();
                resetPositions();
            });

            $scope.$on("orders_updated", function (event, contest) {
                $log.info("[Portfolio] Orders for Contest '" + contest.name + "' updated");
                resetOrders();
            });

            $scope.$on("participant_updated", function (event, contest) {
                $log.info("[Portfolio] Orders for Contest '" + contest.name + "' updated");
                resetPositions();
            });

        }]);

})();

File: app-js/src/main/scala/com/shocktrade/javascript/dashboard/PortfolioController.scala
package com.shocktrade.javascript.dashboard

import org.scalajs.angularjs.Timeout
import org.scalajs.angularjs.cookies.Cookies
import org.scalajs.angularjs.toaster.Toaster
import org.scalajs.nodejs.util.ScalaJsHelper._
import org.scalajs.angularjs.{Controller, Scope, injected}
import com.shocktrade.javascript.AppEvents._
import com.shocktrade.javascript.dialogs.NewOrderDialogController.NewOrderDialogResult
import com.shocktrade.javascript.dialogs.{NewOrderDialog, NewOrderParams}
import com.shocktrade.javascript.discover.DiscoverController
import com.shocktrade.javascript.models._
import com.shocktrade.javascript.{GlobalLoading, MySessionService}
import org.scalajs.dom.console

import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue
import scala.scalajs.js
import scala.util.{Failure, Success}

/**
  * Portfolio Controller
  * @author lawrence.daniels@gmail.com
  */
class PortfolioController($scope: PortfolioScope, $cookies: Cookies, $timeout: Timeout, toaster: Toaster,
                          @injected("MySessionService") mySession: MySessionService,
                          @injected("ContestService") contestService: ContestService,
                          @injected("NewOrderDialog") newOrderDialog: NewOrderDialog)
  extends Controller with GlobalLoading {

  private val marketOrderTypes = js.Array("MARKET", "MARKET_ON_CLOSE")

  $scope.selectedClosedOrder = js.undefined
  $scope.selectedOrder = js.undefined
  $scope.selectedPosition = js.undefined

  $scope.portfolioTabs = js.Array(
    PortfolioTab(name = "Chat", icon = "fa-comment-o", path = "/assets/views/dashboard/chat.htm", active = false),
    PortfolioTab(name = "Positions", icon = "fa-list-alt", path = "/assets/views/dashboard/positions.htm", active = false),
    PortfolioTab(name = "Open Orders", icon = "fa-folder-open-o", path = "/assets/views/dashboard/orders_active.htm", active = false),
    PortfolioTab(name = "Closed Orders", icon = "fa-folder-o", path = "/assets/views/dashboard/orders_closed.htm", active = false),
    PortfolioTab(name = "Performance", icon = "fa-bar-chart-o", path = "/assets/views/dashboard/performance.htm", active = false),
    PortfolioTab(name = "Exposure", icon = "fa-pie-chart", path = "/assets/views/dashboard/exposure.htm", active = false))

  /////////////////////////////////////////////////////////////////////
  //          Closed Order Functions
  /////////////////////////////////////////////////////////////////////

  $scope.getClosedOrders = () => {
    mySession.getClosedOrders.filter(_.accountType == $scope.getAccountType())
  }

  $scope.isClosedOrderSelected = () => {
    $scope.getClosedOrders().nonEmpty && $scope.selectedClosedOrder.nonEmpty
  }

  $scope.selectClosedOrder = (closeOrder: js.UndefOr[ClosedOrder]) => {
    $scope.selectedClosedOrder = closeOrder
  }

  $scope.toggleSelectedClosedOrder = () => {
    $scope.selectedClosedOrder = js.undefined
  }

  /////////////////////////////////////////////////////////////////////
  //          Active Order Functions
  /////////////////////////////////////////////////////////////////////

  $scope.computeOrderCost = (anOrder: js.UndefOr[Order]) => {
    anOrder.map(o => o.price * o.quantity + o.commission)
  }

  $scope.cancelOrder = (aContestId: js.UndefOr[BSONObjectID], aPlayerId: js.UndefOr[BSONObjectID], anOrderId: js.UndefOr[BSONObjectID]) => {
    for {
      contestId <- aContestId
      playerId <- aPlayerId
      orderId <- anOrderId
    } {
      asyncLoading($scope)(contestService.deleteOrder(contestId, playerId, orderId)) onComplete {
        case Success(contest) => mySession.setContest(contest)
        case Failure(err) =>
          toaster.error("Failed to cancel order")
      }
    }
  }

  $scope.getActiveOrders = () => {
    mySession.getOrders filter (_.accountType == $scope.getAccountType())
  }

  $scope.isMarketOrder = (anOrder: js.UndefOr[Order]) => {
    anOrder.exists(order => marketOrderTypes.contains(order.priceType))
  }

  $scope.isOrderSelected = () => $scope.getActiveOrders().nonEmpty && $scope.selectedOrder.nonEmpty

  $scope.popupNewOrderDialog = (anAccountType: js.UndefOr[String]) => anAccountType map { accountType =>
    newOrderDialog.popup(NewOrderParams(
      symbol = $cookies.getOrElse(DiscoverController.LastSymbolCookie, "AAPL"),
      accountType = accountType
    ))
  }

  $scope.selectOrder = (order: js.UndefOr[Order]) => $scope.selectedOrder = order

  $scope.toggleSelectedOrder = () => $scope.selectedOrder = js.undefined

  /////////////////////////////////////////////////////////////////////
  //          Performance Functions
  /////////////////////////////////////////////////////////////////////

  $scope.getPerformance = () => mySession.getPerformance

  $scope.isPerformanceSelected = () => $scope.getPerformance().nonEmpty && $scope.selectedPerformance.nonEmpty

  $scope.selectPerformance = (performance: js.UndefOr[Performance]) => $scope.selectedPerformance = performance

  $scope.toggleSelectedPerformance = () => $scope.selectedPerformance = js.undefined

  $scope.cost = (aTx: js.UndefOr[Performance]) => aTx map { tx =>
    tx.pricePaid * tx.quantity + tx.commissions
  }

  $scope.soldValue = (aTx: js.UndefOr[Performance]) => aTx map { tx =>
    tx.priceSold * tx.quantity
  }

  $scope.proceeds = (tx: js.UndefOr[Performance]) => {
    for {
      soldValue <- $scope.soldValue(tx)
      cost <- $scope.cost(tx)
    } yield soldValue - cost
  }

  $scope.gainLoss = (tx: js.UndefOr[Performance]) => {
    for {
      proceeds <- $scope.proceeds(tx)
      cost <- $scope.cost(tx)
    } yield (proceeds / cost) * 100d
  }

  /////////////////////////////////////////////////////////////////////
  //          Position Functions
  /////////////////////////////////////////////////////////////////////

  $scope.getPositions = () => {
    mySession.participant foreach enrichPositions
    mySession.getPositions filter (_.accountType == $scope.getAccountType())
  }

  $scope.isPositionSelected = () => $scope.getPositions().nonEmpty && $scope.selectedPosition.nonEmpty

  $scope.selectPosition = (position: js.UndefOr[Position]) => $scope.selectedPosition = position

  $scope.sellPosition = (aSymbol: js.UndefOr[String], aQuantity: js.UndefOr[Double]) => {
    for {
      symbol <- aSymbol
      quantity <- aQuantity
    } yield newOrderDialog.popup(NewOrderParams(symbol = symbol, quantity = quantity))
  }

  $scope.toggleSelectedPosition = () => $scope.selectedPosition = js.undefined

  $scope.tradingStart = () => new js.Date()

  /////////////////////////////////////////////////////////////////////
  //          Private Functions
  /////////////////////////////////////////////////////////////////////

  private def enrichOrders(participant: Participant) {
    if (mySession.participant.nonEmpty) {
      if (!isDefined(participant.dynamic.enrichedOrders)) {
        participant.dynamic.enrichedOrders = true

        for {
          contestId <- mySession.getContestID
          playerId <- participant._id
        } {
          contestService.getEnrichedOrders(contestId, playerId) onComplete {
            case Success(enrichedOrders) => mySession.participant.foreach(_.orders = enrichedOrders)
            case Failure(e) =>
              toaster.error("Error!", "Error loading enriched orders")
          }
        }
      }
    }
  }

  private def enrichPositions(participant: Participant) {
    if (mySession.participant.nonEmpty) {
      if (!isDefined(participant.dynamic.enrichedPositions)) {
        participant._id foreach { playerId =>
          participant.dynamic.enrichedPositions = true
          for {
            contestId <- mySession.getContestID
          } {
            contestService.getEnrichedPositions(contestId, playerId) onComplete {
              case Success(enrichedPositions) => mySession.participant.foreach(_.positions = enrichedPositions)
              case Failure(e) => toaster.error("Error loading enriched positions")
            }
          }
        }
      }
    }
  }

  private def resetOrders() {
    $scope.selectedOrder = js.undefined
    $scope.selectedClosedOrder = js.undefined
  }

  private def resetPositions() {
    $scope.selectedPosition = js.undefined
    $scope.selectedPerformance = js.undefined
  }

  //////////////////////////////////////////////////////////////////////
  //              Watch Event Listeners
  //////////////////////////////////////////////////////////////////////

  $scope.$on(ContestSelected, { (event: js.Dynamic, contest: js.Dynamic) =>
    console.log(s"[Portfolio] Contest '${contest.name}' selected")
    resetOrders()
    resetPositions()
  })

  $scope.$on(ContestUpdated, { (event: js.Dynamic, contest: js.Dynamic) =>
    console.log(s"[Portfolio] Contest '${contest.name}' updated")
    resetOrders()
    resetPositions()
  })

  $scope.$on(OrderUpdated, { (event: js.Dynamic, contest: js.Dynamic) =>
    console.log(s"[Portfolio] Orders for Contest '${contest.name}' updated")
    resetOrders()
  })

  $scope.$on(ParticipantUpdated, { (event: js.Dynamic, contest: js.Dynamic) =>
    console.log(s"[Portfolio] Orders for Contest '${contest.name}' updated")
    resetPositions()
  })

}

/**
  * Portfolio Scope
  * @author lawrence.daniels@gmail.com
  */
@js.native
trait PortfolioScope extends Scope {
  // variables
  var portfolioTabs: js.Array[PortfolioTab] = js.native
  var selectedClosedOrder: js.UndefOr[ClosedOrder] = js.native
  var selectedOrder: js.UndefOr[Order] = js.native
  var selectedPerformance: js.UndefOr[Performance] = js.native
  var selectedPosition: js.UndefOr[Position] = js.native

  // closed order functions
  var getClosedOrders: js.Function0[js.Array[ClosedOrder]] = js.native
  var isClosedOrderSelected: js.Function0[Boolean] = js.native
  var selectClosedOrder: js.Function1[js.UndefOr[ClosedOrder], Unit] = js.native
  var toggleSelectedClosedOrder: js.Function0[Unit] = js.native

  // order functions
  var computeOrderCost: js.Function1[js.UndefOr[Order], js.UndefOr[Double]] = js.native
  var cancelOrder: js.Function3[js.UndefOr[BSONObjectID], js.UndefOr[BSONObjectID], js.UndefOr[BSONObjectID], Unit] = js.native
  var getActiveOrders: js.Function0[js.Array[Order]] = js.native
  var getAccountType: js.Function0[String] = js.native
  var isMarketOrder: js.Function1[js.UndefOr[Order], Boolean] = js.native
  var isOrderSelected: js.Function0[Boolean] = js.native
  var selectOrder: js.Function1[js.UndefOr[Order], Unit] = js.native
  var popupNewOrderDialog: js.Function1[js.UndefOr[String], js.UndefOr[js.Promise[NewOrderDialogResult]]] = js.native
  var toggleSelectedOrder: js.Function0[Unit] = js.native

  // performance functions
  var cost: js.Function1[js.UndefOr[Performance], js.UndefOr[Double]] = js.native
  var gainLoss: js.Function1[js.UndefOr[Performance], js.UndefOr[Double]] = js.native
  var proceeds: js.Function1[js.UndefOr[Performance], js.UndefOr[Double]] = js.native
  var soldValue: js.Function1[js.UndefOr[Performance], js.UndefOr[Double]] = js.native
  var getPerformance: js.Function0[js.Array[Performance]] = js.native
  var isPerformanceSelected: js.Function0[Boolean] = js.native
  var selectPerformance: js.Function1[js.UndefOr[Performance], Unit] = js.native
  var toggleSelectedPerformance: js.Function0[Unit] = js.native

  // position functions
  var getPositions: js.Function0[js.Array[Position]] = js.native
  var isPositionSelected: js.Function0[Boolean] = js.native
  var selectPosition: js.Function1[js.UndefOr[Position], Unit] = js.native
  var sellPosition: js.Function2[js.UndefOr[String], js.UndefOr[Double], js.UndefOr[js.Promise[NewOrderDialogResult]]] = js.native
  var toggleSelectedPosition: js.Function0[Unit] = js.native
  var tradingStart: js.Function0[js.Date] = js.native

}

/**
  * Portfolio Tab
  * @author lawrence.daniels@gmail.com
  */
@js.native
trait PortfolioTab extends js.Object {
  var name: String = js.native
  var icon: String = js.native
  var path: String = js.native
  var active: Boolean = js.native
}

/**
  * Portfolio Tab Companion Object
  * @author lawrence.daniels@gmail.com
  */
object PortfolioTab {

  def apply(name: String, icon: String, path: String, active: Boolean = false) = {
    val tab = New[PortfolioTab]
    tab.name = name
    tab.icon = icon
    tab.path = path
    tab.active = active
    tab
  }

}

